File: grind75/cpp/week1/050-Course-Schedule-Medium.cpp
// Problem link: https://leetcode.com/problems/course-schedule/description/

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // Edge case
        if (prerequisites.empty()) {
            return true;
        }

        unordered_set<int> visited;
        unordered_set<int> processed;
        unordered_map<int, vector<int>> courseGraph;

        // Build adjacency list
        for (const auto& prereq : prerequisites) {
            courseGraph[prereq[0]].push_back(prereq[1]);
        }

        function<bool(int)> dfs = [&](int course) {
            // Check if the course is already visited (cycle detected)
            if (visited.find(course) != visited.end()) {
                return false;
            }

            // Check if the course is already processed (no need to revisit)
            if (processed.find(course) != processed.end()) {
                return true;
            }

            visited.insert(course);
            for (int prereq : courseGraph[course]) {
                if (!dfs(prereq)) {
                    return false;
                }
            }

            visited.erase(course);
            processed.insert(course);

            return true;
        };

        for (int course = 0; course < numCourses; ++course) {
            if (!dfs(course)) {
                return false;
            }
        }

        return true;
        
    }
};

File: grind75/python/week1/050-Course-Schedule-Medium.py
# Question link: https://leetcode.com/problems/course-schedule/

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        # edge case
        if not prerequisites:
            return True
        
        visited = set()
        processed = set()
        courseGraph = defaultdict(list)

        # build out our adjency lists
        for course, prereq in prerequisites:
            courseGraph[course].append(prereq)
        
        def dfs(course):

            # Check if the course we are on already has been visited
            if course in visited:
                return False
            
            # Check if course we are on has already been processed
            # This is to skip redundant work since we know we can fulfill it's prereqs already
            if course in processed:
                return True

            visited.add(course)
            for prereq in courseGraph[course]:
                if not dfs(prereq):
                    return False

            visited.remove(course)
            processed.add(course)

            return True
        
        for course in range(numCourses):
            if not dfs(course):
                return False
        
        return True

        # 0, 1, 2
